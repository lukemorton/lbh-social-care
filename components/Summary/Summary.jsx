import PropTypes from 'prop-types';
import Link from 'next/link';

import SummaryList from 'components/Summary/SummaryList';
import { filterStepsOnCondition, filterDataOnCondition } from 'utils/steps';

const MultiValue = (value) => (
  <div key={value}>
    <span>{value}</span>
    <br />
  </div>
);

const SummaryMultiSection = ({
  formData,
  id,
  title,
  canEdit,
  formPath,
  ...props
}) => (
  <>
    {formData[id].map((data, key) => (
      <SummarySection
        {...props}
        canEdit={canEdit}
        formData={data}
        formPath={formPath}
        title={`${title} - ${key + 1}`}
        key={`${id}/${key}`}
        id={`${id}/${key + 1}`}
      />
    ))}
    {canEdit && (
      <p className="govuk-!-margin-bottom-7">
        <Link
          href={`${formPath}${id}/${formData[id].length + 1}?fromSummary=true`}
          as={`${formPath}${id}/${formData[id].length + 1}?fromSummary=true`}
        >
          <a className="govuk-link">Add Another {title}</a>
        </Link>
      </p>
    )}
  </>
);

export const SummarySection = ({
  formData,
  id,
  title,
  components,
  formPath,
  canEdit,
}) => {
  const Summary = (
    <SummaryList
      list={components
        .filter(({ name }) => formData[name])
        .map(({ name, label }) => ({
          key: name,
          title: label,
          value: Array.isArray(formData[name])
            ? formData[name]
                .filter(Boolean)
                .map((v) => MultiValue(v.split('/').pop()))
            : typeof formData[name] === 'object'
            ? Object.values(formData[name]).filter(Boolean).map(MultiValue)
            : typeof formData[name] === 'boolean'
            ? JSON.stringify(formData[name])
            : formData[name],
        }))}
    />
  );
  return (
    <div className="govuk-!-margin-bottom-7">
      <h3 className="govuk-heading-m">{title}</h3>
      {Summary}
      {canEdit && (
        <Link
          href={`${formPath}${id}?fromSummary=true`}
          as={`${formPath}${id}?fromSummary=true`}
        >
          <a className="govuk-link">Change</a>
        </Link>
      )}
    </div>
  );
};

const Summary = ({ formData, formSteps, formPath, canEdit }) =>
  filterStepsOnCondition(formSteps, formData).map((section) => {
    const props = {
      key: section.id,
      formData: filterDataOnCondition(formSteps, formData),
      formPath: formPath,
      canEdit: canEdit,
      ...section,
    };
    return section.isMulti ? (
      <SummaryMultiSection {...props} />
    ) : (
      <SummarySection {...props} />
    );
  });

Summary.propTypes = {
  formData: PropTypes.shape({}).isRequired,
  formSteps: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  formPath: PropTypes.string.isRequired,
  canEdit: PropTypes.bool,
};

export default Summary;
