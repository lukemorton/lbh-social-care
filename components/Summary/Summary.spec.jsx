import { render } from '@testing-library/react';

import Summary from './Summary';

describe('Summary component', () => {
  const props = {
    formData: {
      foo: '123',
      bar: 'asd',
      asd: '',
      qwe: 'yo',
      foobar: 'asd',
    },
    formPath: '/form/foo/',
    formSteps: [
      {
        id: 'first-step',
        title: 'First Step',
        components: [
          { component: 'TextInput', label: 'I am foo component', name: 'foo' },
          { component: 'TextInput', label: 'I am bar component', name: 'bar' },
          { component: 'TextInput', label: 'I am asd component', name: 'asd' },
        ],
      },
      {
        id: 'second-step',
        title: 'Second Step',
        components: [
          {
            component: 'TextInput',
            label: 'I am foobar component',
            name: 'foobar',
          },
        ],
      },
    ],
    onFormSubmit: jest.fn(),
  };

  it('should render properly', () => {
    const { asFragment } = render(<Summary {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
