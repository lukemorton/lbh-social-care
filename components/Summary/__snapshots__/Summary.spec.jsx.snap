// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary component should render properly 1`] = `
<DocumentFragment>
  <div
    class="govuk-!-margin-bottom-7"
  >
    <h3
      class="govuk-heading-m"
    >
      First Step
    </h3>
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          I am foo component
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          123
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          I am bar component
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          asd
        </dd>
      </div>
    </dl>
  </div>
  <div
    class="govuk-!-margin-bottom-7"
  >
    <h3
      class="govuk-heading-m"
    >
      Second Step
    </h3>
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          I am foobar component
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          asd
        </dd>
      </div>
    </dl>
  </div>
</DocumentFragment>
`;
